Plotting module - Backend support

Davide-sd opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

Fixes #20113 .

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

Previously, the keyword `backend` accepted only a string value. Now, it
accepts also a subclass of BaseBackend. This allows users to define their
custom backends (for example, creating a backend for Plotly, Bokeh, ...).
Without this edit, the user would be forced to subclass Plot, PlotGrid,
Basebackedn as well as re-creating a function similar to plot.

Example:

```
class DummyBackend(BaseBackend):
    def show(self):
        pass
x = symbols("x")
plot(sin(x), backend=DummyBackend)
```


#### Other comments

This PR was originally part of PR 20114. Requested separation for better clarity.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
Does the docstring for `BaseBackend` appear in the html docs anywhere?

Davide-sd commented:
@oscarbenjamin now it should.

oscarbenjamin commented:
The HTML build seems to have failed:
```
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.BaseBackend:14: WARNING: Inline emphasis start-string without end-string.
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.BaseBackend:17: WARNING: Inline emphasis start-string without end-string.
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.BaseBackend:24: WARNING: Inline emphasis start-string without end-string.
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.BaseBackend:24: WARNING: Inline emphasis start-string without end-string.
```

Davide-sd commented:
Done.

oscarbenjamin committed:
Plotting - Improving backend support + Plot size

1. Previously, the keyword `backend` accepted only a string value. Now, it
accepts also a subclass of BaseBackend. This allows users to define their
custom backends (for example, creating a backend for Plotly, Bokeh, ...).
Without this edit, the user would be forced to subclass Plot, PlotGrid,
Basebackedn as well as re-creating a function similar to plot.

Example:
class DummyBackend(BaseBackend):
    def show(self):
        pass
x = symbols("x")
plot(sin(x), backend=DummyBackend)

2. Added the keyword argument `size` to Plot, PlotGrid and plot functions.
The user can now specify the figure size. This is useful for generating
plots that are ready to be exported and later included in documents.
Obviously, this only works with `MatplotlibBackend`.

Example:
x = symbols("x")
plot(sin(x), size=(8, 4))

Co-authored-by: Davide_sd <sandona.davide@gmail.com>

oscarbenjamin commented:
There were commits adding and removing a folder called `.directory`. I've squashed the commits here to remove those.

oscarbenjamin commented:
I've also opened #20463 which is the same as this but for the 1.7 release branch

oscarbenjamin commented:
Thanks @Davide-sd!
